catalog/IndexScheme.hpp

```diff
@@ -0,0 +1,195 @@
+   * @return Whether proto is fully-formed and valid.
+   **/
+  static bool ProtoIsValid(const serialization::IndexScheme &proto);
+
+  /**
+   * @brief Serialize the Index Scheme as Protocol Buffer.
+   *
+   * @return The Protocol Buffer representation of Index Scheme.
+   **/
+  serialization::IndexScheme getProto() const;
+
+  /**
+   * @brief Get the number of indices for the relation.
+   *
+   * @return The number of indices defined for the relation.
+   **/
+  inline const std::size_t getNumIndices() const {
+    return index_map_.size();
+  }
+
+  /**
+   * @brief Check whether an index with the given exists or not.
+   *
+   * @param index_name Name of the index to be checked.
+   * @return Whether the index exists or not.
+   **/
+  bool hasIndexWithName(const std::string &index_name) const {
+    return index_map_.find(index_name) != index_map_.end();
+  }
+
+  /**
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
+   *
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map.
+    for (auto cit = index_map_.cbegin(); cit != index_map_.cend(); ++cit) {
+      const std::vector<IndexSubBlockDescription> &index_descriptions = cit->second;
+      for (auto index_description_it = index_descriptions.cbegin();
+           index_description_it != index_descriptions.cend();
+           ++index_description_it) {
+        // Check if the stored description matches the given description.
+        if (areIndexDescriptionsSame(*index_description_it, index_description)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @brief Check whether two index descriptions are same or not.
+   *        Two index descriptions are same if they have any matching
+   *        attributes ids and same index type.
+   *
+   * @param description_expected First index description.
+   * @param description_checked Second index description.
+   * @return Whether the two index_descriptions are similar or not.
+   **/
+  bool areIndexDescriptionsSame(const IndexSubBlockDescription &description_expected,
+                                const IndexSubBlockDescription &description_checked) const {
+    if (description_expected.sub_block_type() != description_checked.sub_block_type()) {
+      return false;
+    }
+
+    // Check if there are matching attribute ids for both indices of same type.
+    // Create a hash map of attribute ids in description_expected,
+    // and probe the hash map for duplicates using the attribute ids from description_checked.
+    std::unordered_map<std::int32_t, bool> existence_map;
```

Comment:
Why need the second `bool` part here? I think an `unordered_set` of `attribute_id`s is good enough to find the duplicates from `description_checked`.